### PR TITLE
[WIP] Toy to RISCV compiler + emulator

### PR DESCRIPTION
**Pull request by superlopuh**
Toy to RISC-V compiler that leverages built-in dialects + RISC-V, implemented in xDSL. Optionally compile to custom L1 accelerator instructions with custom implementations in `riscemu`. Pure Python.

TODO:

- [ ] remove mlir-opt use
    - [ ] affine to scf
    - [ ] scf to cf
- [x] remove vector dialect
- [x] affine to accelerator
- [x] scf to riscemu bug (probably due to my implementation of the lowering)
- [x] cf to riscemu bug (riscemu j instruction jumps to absolute address not relative offset)
- [ ] merge all dialect changes in xDSL
- [ ] merge all interpreter function impls in xDSL

### CODE REVIEW

File: xdsl/dialects/riscv.py [hunk 1]
```diff
@@ -704,6 +1052,321 @@ class LiOp(RdImmOperation):
     name = "riscv.li"
 
 
+@irdl_op_definition
+class LBOp(RdRsImmOperation):
+    name = "riscv.lb"
```
Comment (tobiasgrosser) [inline xdsl/dialects/riscv.py:1057]: This and the obs below are already upstream. I feel the rebase / merge accidentally copied these operations. Maybe drop the duplicated ones to make the diff smaller.

Comment (superlopuh) [reply xdsl/dialects/riscv.py:1057]: Indeed, will drop

**Comment by tobiasgrosser** [comment]
@superlopuh, after merging #1019, this PR seems to have little left. Are we close to merging it?

**Comment by superlopuh** [comment]
I added the TODO to the PR description, I'll try to keep it up to date as we go. The riscemu PR is already up, we just need a riscemu version so we're not pointing to a commit on my branch over there. I need to write tests for the LLVM dialect changes and open up the PR for that, has been low priority recently. We also need to discuss what flow we want for this tutorial. Maybe it doesn't matter too much, and we can just merge what I have right now, or maybe we can keep going with my initial push of 1:1 filecheck compatibility with MLIR Toy as long as possible. I didn't want to discuss it while I was still actively working on dependencies, but it's close to being ready.

**Comment by superlopuh** [comment]
This pretty much all works already. With the merge of the assembly printing of loops PR, the only thing remaining will be to add a line to the filecheck test. I'll do that in an upcoming PR. I have a wish list of changes to make, such as removing the dependency on `mlir-opt`, and on all riscv-related code in the Toy folder, as the functionality should be in the xDSL framework itself, but that can all be done in upcoming PRs also.

**Pull request closed without merging**